File: JavaScript/791-Custom Sort String.js
/**
 * @param {string} order
 * @param {string} str
 * @return {string}
 */
var customSortString = function(order, str) {
    let hmap = new Map();
    const N = order.length;
    for (let i=0; i<N; i++){
        hmap.set(order[i], N-i);
    }
    let res = Array.from(str).sort(function(a, b){
        if (hmap.has(a) && hmap.has(b)){
            return hmap.get(b) - hmap.get(a);
        } else if(hmap.has(a)){
            return -1;
        }else if(hmap.has(b)) {
            return 1;
        }else{
            return 0;
        }
    })
    return res.join("");
};

File: C++/791-Custom Sort String.cpp
class Solution {
public:
    string customSortString(string order, string s) {
        unordered_map<char, int> hmap;
        
        for (int i=0; i<order.size(); i++) {
            hmap[order[i]] = i;
        }
        
        sort(s.begin(), s.end(), [&](char a, char b){ return hmap[a] < hmap[b];});
        
        return s;
    }
};